### ISSUE TITLE
GoPro Hero 11 Black

### ISSUE THREAD

**Issue by reox** [lensfun/lensfun#2246]
I obtained a GoPro Hero 11 Black and would also like to provide calibration data.

I read in #2096 that there are some other requirements (especially this comment: https://github.com/lensfun/lensfun/issues/2096#issuecomment-1686503195) but I'm not sure where I would find such buildings...
Could @tuxfanx or @pengowray maybe point me to some photos that were used in the calibration as a reference?

Are there any hints for the vignetting? Does it work in the same way as for regular lenses? I have obtained some material from a previous lens calibration to get a reasonable diffuse light, but as I cannot set the focus on the lens, how is this done?

**Comment by pengowray** [comment]
Hi @reox , I'm not sure which photo(s) were used in the end for the Hero 10, or if any were mine. I did try to follow instructions but even some of largest building in the CBD here didn't stretch across enough of the field of view. Later I did see someone else's example photos for a very wide lens calibration which had buildings stretching across most of the image. Rather than going for the large tall buildings, they were **large long buildings, probably at a university campus**, so that could be the sort of place where if you get close enough to a building you might get some good long horizontal lines.

I didn't attempt to take any photos for vignetting calibration so I can't comment on that. I did wonder how you could go about it with a GoPro or fisheye lens.

Hope that helps a little and hopefully someone else can chime in.

**Comment by reox** [comment]
Thanks @pengowray - I know some long buildings (I already calibrated another lens using it) but they have discontinuous lines... I may give it a try though!

I went over the instructions again and another thing caught my attention: There is no way to control for ISO in the gopro, right? Is that a problem? And I guess the focus is fixed in the lens?

**Comment by pengowray** [comment]
Yes, you can control ISO in photo mode on the GoPro Hero 10 so I'd assume you can on the 11. You can also shoot RAW, which would be recommended for this. 

Might also want to use the 3s timer if it helps keep the shot steady if you're resting the GoPro on something or have it on a tripod, though in full sun it's probably not really needed.

The lens focus is fixed, yes. Wide angle lens and smaller sensors both give greater Depth of Field. GoPros have both, so shots generally appear in focus everywhere.

**Comment by reox** [comment]
Ahh it is a bit hidden with min and max iso - but true, you can force it.

I just had another idea of using the jpgs to get the coefficients by finding controlpoints via `cpfind`:

```
<distortion model="ptlens" focal="2.7" a="-0.085288" b="0.154603" c="-0.072776" />
```

Here is my mediocre script to generate that:

```
import matplotlib.pyplot as plt
import numpy as np
from scipy.optimize import curve_fit
import os
import glob

#   Imagefile: GOPR0114.tif      ---> RAW
#   Imagefile: GOPR0114_01.tif   ---> JPEG

# First run pto_gen:
#   pto_gen GOPR0114.tif GOPR0114_01.tif -o input.pto

# Then run cpfind:
#   cpfind input.pto -o foobaz.pto --sieve1width 100 --sieve1height 100 --sieve2width 10 --sieve2height 10 
# to generate some more points than the default 25, some settings were adjusted

for raw in glob.glob("GOPR????.tif"):
    f, e = os.path.splitext(raw)
    jpeg = f + "_01" + e
    os.system(f"pto_gen {raw} {jpeg} -o input.pto")
    os.system(f"cpfind input.pto -o {f}.pto --sieve1width 100 --sieve1height 100 --sieve2width 10 --sieve2height 10")

# Read control points
cp = np.array(
    [
        [float(z[1:]) for z in x.split(" ")[3:7]]
        for ptofile in glob.glob("GOPR????.pto")
        for x in open(ptofile).read().splitlines()
        if x.startswith('c ') and not x.endswith("t1")
    ]
)

print(f"Found {len(cp)} control points")
nx, ny = 5568, 4872  # Size of image in pixels

# Coordinates
x, y, X, Y = cp.T

# to origin at 0/0
x = x - (nx / 2)
y = y - (ny / 2)
X = X - (nx / 2)
Y = Y - (ny / 2)

# to unit radius by height
x = x / (ny // 2)
y = y / (ny // 2)
X = X / (ny // 2)
Y = Y / (ny // 2)

# Calculate radius
R = np.sqrt(x**2 + y**2)
r = np.sqrt(X**2 + Y**2)

def distortion(r, a, b, c):
    d = 1 - a - b - c
    return r * (a * r**3 + b * r**2 + c * r + d)

(a, b, c), pcov, infodict, msg, ier = curve_fit(distortion, r, R, full_output=True)

print("Mean Residuum", infodict["fvec"].mean())
print(msg)
print()
print(f'<distortion model="ptlens" focal="2.7" a="{a:.6f}" b="{b:.6f}" c="{c:.6f}" />')

plt.scatter(x, y, label='raw')
plt.scatter(X, Y, marker='*', label='jpg')
plt.gca().add_patch(plt.Circle((0, 0), 1, edgecolor='black', facecolor='none', label='non-scaling restraint'))
plt.gca().add_patch(plt.Rectangle((-nx / ny, -1), 2*(nx/ny), 2, edgecolor='black', facecolor='none', linestyle='--'))
plt.axhline(0, linestyle=':', color='black')
plt.axvline(0, linestyle=':', color='black')
plt.axis('equal')
plt.legend()
plt.title(f'Fit: a={a:.5f}, b={b:.5f}, c={c:.5f}')
plt.show()
```

It gets close but is not the same... Maybe the distortion model should be another one?
